In [144]:
import pandas as pd

interpro = pd.read_csv('interpro_results.csv')
interpro.fillna('N/A', inplace=True)
interpro


species_name         prot_id  \
0       kluyveromyces_lactis_cds.fna     XP_451741.1   
1       kluyveromyces_lactis_cds.fna     XP_451741.1   
2       kluyveromyces_lactis_cds.fna     XP_451741.1   
3       kluyveromyces_lactis_cds.fna     XP_451741.1   
4       kluyveromyces_lactis_cds.fna     XP_453196.1   
...                              ...             ...   
8688  aplosporella_prunicola_cds.fna  XP_033397713.1   
8689  aplosporella_prunicola_cds.fna  XP_033401083.1   
8690  aplosporella_prunicola_cds.fna  XP_033401625.1   
8691  aplosporella_prunicola_cds.fna  XP_033402011.1   
8692  aplosporella_prunicola_cds.fna  XP_033402633.1   

                                         gene_id  \
0            lcl|NC_006038.1_cds_XP_451741.1_696   
1            lcl|NC_006038.1_cds_XP_451741.1_696   
2            lcl|NC_006038.1_cds_XP_451741.1_696   
3            lcl|NC_006038.1_cds_XP_451741.1_696   
4           lcl|NC_006040.1_cds_XP_453196.1_2079   
...                                          ...   
8688  lcl|NW_022983540.1_cds_XP_033397713.1_5413   
8689  lcl|NW_022983532.1_cds_XP_033401083.1_1800   
8690  lcl|NW_022983531.1_cds_XP_033401625.1_1234   
8691  lcl|NW_022983531.1_cds_XP_033402011.1_1622   
8692  lcl|NW_022983530.1_cds_XP_033402633.1_1120   

                 biological_process_go  \
0                                  N/A   
1                                  N/A   
2                                  N/A   
3                                  N/A   
4                                  N/A   
...                                ...   
8688                               N/A   
8689  GO:0000278;GO:0000226;GO:0007017   
8690                               N/A   
8691                               N/A   
8692                               N/A   

                             biological_process_go_name  \
0                                                   N/A   
1                                                   N/A   
2                                                   N/A   
3                                                   N/A   
4                                                   N/A   
...                                                 ...   
8688                                                N/A   
8689  mitotic cell cycle;microtubule cytoskeleton or...   
8690                                                N/A   
8691                                                N/A   
8692                                                N/A   

                 molecular_function_go  \
0                           GO:0003755   
1                           GO:0003755   
2                           GO:0003755   
3                           GO:0003755   
4                GO:0061133;GO:0070628   
...                                ...   
8688                               N/A   
8689  GO:0005525;GO:0003924;GO:0005200   
8690                               N/A   
8691                               N/A   
8692             GO:0016791;GO:0003993   

                             molecular_function_go_name  \
0          peptidyl-prolyl cis-trans isomerase activity   
1          peptidyl-prolyl cis-trans isomerase activity   
2          peptidyl-prolyl cis-trans isomerase activity   
3          peptidyl-prolyl cis-trans isomerase activity   
4     endopeptidase activator activity;proteasome bi...   
...                                                 ...   
8688                                                N/A   
8689  GTP binding;GTPase activity;structural constit...   
8690                                                N/A   
8691                                                N/A   
8692     phosphatase activity;acid phosphatase activity   

                 cellular_component_go  \
0                GO:0000785;GO:0005730   
1                GO:0000785;GO:0005730   
2                GO:0000785;GO:0005730   
3                GO:0000785;GO:0005730   
4     GO:0008541;GO:0005634;GO:0005737   
...                                ...   
8688

In [145]:
guides = interpro['targeting_guide'].unique().tolist()
guides

['AGAGGAAGAGGAAGAGGAAG',
 'GCTGGTGACACTCACTTGGG',
 'AGAAGAGGAGGAGGAGGAGG',
 'ATCCAATGGAGATTCCAAAC',
 'GCTTTCTTGCATTGGTACAC',
 'GAGGCTGGTATCTCCAAGGA',
 'TGCTGTTGCTGTTGCTGCTG',
 'TCTTCATCGGCTTCGGCGAA']

In [146]:
interpro[interpro['targeting_guide'] == 'GCTTTCTTGCATTGGTACAC'][['species_name', 'prot_id', 'family_locations', 'guide_nuc_start_position', 'prot_ranges']]

species_name             prot_id  \
17     trichodelitschia_bisporula_cds.fna        KAF2401180.1   
60           aspergillus_fischeri_cds.fna      XP_001264678.1   
76     nemania_maritima_azju_v1_0_cds.fna  fgenesh1_pm.5_#_35   
78                nemania_serpens_cds.fna        KAI1195749.1   
86     aspergillus_pseudocaelatus_cds.fna        KAE8416686.1   
...                                   ...                 ...   
8647        xylographa_trunciseda_cds.fna        MCJ1401122.1   
8649  colletotrichum_higginsianum_cds.fna      XP_018151417.1   
8661     trichoderma_arundinaceum_cds.fna            RFU78125   
8668    annulohypoxylon_moriforme_cds.fna        KAI1457021.1   
8689       aplosporella_prunicola_cds.fna      XP_033401083.1   

                                       family_locations  \
17    41-58|88-99|108-126|152-164|212-224|230-239|24...   
60    41-58|88-99|108-126|152-164|212-224|230-239|24...   
76    1-428;347-370|311-329|41-58|212-224|265-281|37...   
78    41-58|88-99|108-126|152-164|212-224|230-239|24...   
86    10-30|51-70|93-104|106-130|132-150|151-172|176...   
...                                                 ...   
8647  151-172|106-130|190-210|132-150|370-398|176-18...   
8649  41-58|88-99|108-126|152-164|212-224|230-239|24...   
8661  41-58|88-99|108-126|152-164|212-224|230-239|24...   
8668  10-30|51-70|93-104|106-130|132-150|151-172|176...   
8689  41-58|88-99|108-126|152-164|212-224|230-239|24...   

      guide_nuc_start_position prot_ranges  
17                        1177     393-400  
60                        1177     393-400  
76                        1177     393-400  
78                        1177     393-400  
86                        1177     393-400  
...                        ...         ...  
8647                      1177     393-400  
8649                      1177     393-400  
8661                      1177     393-400  
8668                      1177     393-400  
8689                      1177     393-400  

[558 rows x 5 columns]

In [166]:
def parse_names(s):
    return s.split(';')

def parse_ranges(ranges_str):
    ranges = list()
    for range_group in ranges_str.split(';'):
        if '|' in range_group:
            sub_ranges = list()
            for r in range_group.split('|'):
                start, end = map(int, r.split('-'))
                sub_ranges.append((start, end))
            ranges.append(sub_ranges)
        else:
            start, end = map(int, range_group.split('-'))
            ranges.append([(start, end)])
        
    return ranges

def find_ranges_for_prot_range(ranges, prot_range):
    prot_start, prot_end = map(int, prot_range.split('-'))
    matching_idx = list()
    for idx, rlist in enumerate(ranges):
        for start, end in rlist:
            # Check if prot_range overlaps with the range
            if prot_start >= start and prot_end <= end:
                matching_idx.append(idx)
    return matching_idx
    
def find_target_families(row):
    locs_str = row['family_locations']
    
    if locs_str == 'N/A':
        return []

    names = parse_names(row['families'])
    acc = parse_names(row['family_accessions'])
    ranges = parse_ranges(row['family_locations'])

    prot_range = row['prot_ranges']
    matching_idx = find_ranges_for_prot_range(ranges, prot_range)
    
    return list(set([f'{names[idx]} ({acc[idx]})' for idx in matching_idx]))

def find_target_domains(row):
    locs_str = row['domain_locations']
    
    if locs_str == 'N/A':
        return []

    names = parse_names(row['domains'])
    acc = parse_names(row['domain_accessions'])
    ranges = parse_ranges(row['domain_locations'])

    prot_range = row['prot_ranges']
    matching_idx = find_ranges_for_prot_range(ranges, prot_range)
    
    return list(set([f'{names[idx]} ({acc[idx]})' for idx in matching_idx]))

def find_target_superfamilies(row):
    locs_str = row['superfamily_locations']
    
    if locs_str == 'N/A':
        return []

    names = parse_names(row['superfamilies'])
    acc = parse_names(row['superfamily_accessions'])
    ranges = parse_ranges(row['superfamily_locations'])

    prot_range = row['prot_ranges']
    matching_idx = find_ranges_for_prot_range(ranges, prot_range)
    
    return list(set([f'{names[idx]} ({acc[idx]})' for idx in matching_idx]))

def find_target_conserved_sites(row):
    locs_str = row['conserved_site_locations']
    
    if locs_str == 'N/A':
        return []

    names = parse_names(row['conserved_sites'])
    acc = parse_names(row['conserved_site_accessions'])
    ranges = parse_ranges(row['conserved_site_locations'])

    prot_range = row['prot_ranges']
    matching_idx = find_ranges_for_prot_range(ranges, prot_range)
    
    return list(set([f'{names[idx]} ({acc[idx]})' for idx in matching_idx]))

def find_target_repeats(row):
    locs_str = row['repeat_locations']
    
    if locs_str == 'N/A':
        return []

    names = parse_names(row['repeats'])
    acc = parse_names(row['repeat_accessions'])
    ranges = parse_ranges(row['repeat_locations'])

    prot_range = row['prot_ranges']
    matching_idx = find_ranges_for_prot_range(ranges, prot_range)
    
    return list(set([f'{names[idx]} ({acc[idx]})' for idx in matching_idx]))

def find_target_active_sites(row):
    locs_str = row['active_site_locations']
    
    if locs_str == 'N/A':
        return []

    names = parse_names(row['active_sites'])
    acc = parse_names(row['active_site_accessions'])
    ranges = parse_ranges(locs_str)

    prot_range = row['prot_ranges']
    matching_idx = find_ranges_for_prot_range(ranges, prot_range)
    
    return list(set([f'{names[idx]} ({acc[idx]})' for idx in matching_idx]))

def find_target_binding_sites(row):
    locs_str = row['binding_site_locations']
    
    if locs_str == 'N/A':
        return []

    names = parse_names(row['binding_sites'])
    acc = parse_names(row['binding_site_accessions'])
    ranges = parse_ranges(locs_str)

    prot_range = row['prot_ranges']
    matching_idx = find_ranges_for_prot_range(ranges, prot_range)
    
    return list(set([f'{names[idx]} ({acc[idx]})' for idx in matching_idx]))

def find_target_other_features(row):
    locs_str = row['other_feature_locations']
    
    if locs_str == 'N/A':
        return []

    names = parse_names(row['other_features'])
    acc = parse_names(row['other_feature_accessions'])
    ranges = parse_ranges(locs_str)

    prot_range = row['prot_ranges']
    matching_idx = find_ranges_for_prot_range(ranges, prot_range)
    
    return list(set([f'{names[idx]} ({acc[idx]})' for idx in matching_idx]))



In [148]:
interpro.apply(find_target_families, axis=1).value_counts()

[]                                                                            5440
[Transl_elong_EF1A_euk/arc (IPR004539), TRAFAC_GTPase_members (IPR050100)]     725
[Tubulin (IPR000217)]                                                          555
[Hsp_70_fam (IPR013126)]                                                       368
[ABCF_EF3_subfamily (IPR050611)]                                                48
                                                                              ... 
[TAF7 (IPR037817)]                                                               1
[Sec62 (IPR004728)]                                                              1
[tRNA_Ile_lys_synt (IPR012094)]                                                  1
[Gfo/Idh/MocA_Oxidoreductases (IPR051450)]                                       1
[His_Pase_clade-2 (IPR000560), Histidine_acid_Pase_euk (IPR016274)]              1
Name: count, Length: 507, dtype: int64

In [149]:
interpro.apply(find_target_domains, axis=1).value_counts()

[]                                                                    7593
[T_Tr_GTP-bd_dom (IPR000795)]                                          735
[REJ_dom (IPR014010)]                                                   58
[ABC_transporter-like_ATP-bd (IPR003439), AAA+_ATPase (IPR003593)]      55
[BIP_NBD (IPR042050)]                                                   47
                                                                      ... 
[SSU_processome_Utp12 (IPR007148)]                                       1
[DNA2/NAM7-like_C (IPR041679), SF1_C_Upf1 (IPR047187)]                   1
[Sulf_transp (IPR007272)]                                                1
[DUS-like_FMN-bd (IPR035587)]                                            1
[Cation/H_antiporter_C (IPR013928)]                                      1
Name: count, Length: 98, dtype: int64

In [150]:
interpro.apply(find_target_superfamilies, axis=1).value_counts()

[]                                                 6448
[P-loop_NTPase (IPR027417)]                         774
[Tub_FtsZ_C (IPR008280), Tubulin_C (IPR023123)]     557
[ATPase_NBD (IPR043129)]                            367
[ARM-type_fold (IPR016024)]                          51
                                                   ... 
[PLC-like_Pdiesterase_TIM-brl (IPR017946)]            1
[RMI1_N_C_sf (IPR042470)]                             1
[IPNS-like_sf (IPR027443)]                            1
[C2_domain_sf (IPR035892)]                            1
[Cyt_P450_sf (IPR036396)]                             1
Name: count, Length: 121, dtype: int64

In [151]:
interpro[interpro['targeting_guide'] == 'GCTTTCTTGCATTGGTACAC'].apply(find_target_families, axis=1).value_counts()

[Tubulin (IPR000217)]    555
[]                         3
Name: count, dtype: int64

# [] MEANS THERE ARE 3 PROTEINS WHERE THIS GUIDE DOESN'T TARGET ANY SUPERFAMILIES IN ITS RANGE. It does not mean that there are no superfamilies in that protein.

In [152]:
interpro[interpro['targeting_guide'] == 'GCTTTCTTGCATTGGTACAC'].apply(find_target_domains, axis=1).value_counts()

[]    558
Name: count, dtype: int64

In [153]:
interpro[interpro['targeting_guide'] == 'GCTTTCTTGCATTGGTACAC'].apply(find_target_superfamilies, axis=1).value_counts()

[Tub_FtsZ_C (IPR008280), Tubulin_C (IPR023123)]    557
[]                                                   1
Name: count, dtype: int64

In [154]:
interpro[interpro['targeting_guide'] == 'GCTTTCTTGCATTGGTACAC'].apply(find_target_conserved_sites, axis=1).value_counts()

[]    558
Name: count, dtype: int64

In [155]:
interpro[interpro['targeting_guide'] == 'GCTTTCTTGCATTGGTACAC'].apply(find_target_repeats, axis=1).value_counts()

[]    558
Name: count, dtype: int64

In [156]:
interpro[interpro['targeting_guide'] == 'GCTTTCTTGCATTGGTACAC'].apply(find_target_active_sites, axis=1).value_counts()

[]    558
Name: count, dtype: int64

In [157]:
interpro[interpro['targeting_guide'] == 'GCTTTCTTGCATTGGTACAC'].apply(find_target_other_features, axis=1).value_counts()

[beta_tubulin (cd02187)]                                                     545
[beta_tubulin (cd02187), Cytoplasmic domain (CYTOPLASMIC_DOMAIN)]              9
[Non cytoplasmic domain (NON_CYTOPLASMIC_DOMAIN), beta_tubulin (cd02187)]      3
[]                                                                             1
Name: count, dtype: int64

# Guide GAGGCTGGTATCTCCAAGGA

In [161]:
interpro[interpro['targeting_guide'] == 'GAGGCTGGTATCTCCAAGGA'].apply(find_target_domains, axis=1).value_counts()

[T_Tr_GTP-bd_dom (IPR000795)]    735
[]                                 2
Name: count, dtype: int64

There are 735/737 proteins where this guide cuts in the T_Tr_GTP-bd_dom (IPR000795) domain.

In [162]:
interpro[interpro['targeting_guide'] == 'GAGGCTGGTATCTCCAAGGA'].apply(find_target_families, axis=1).value_counts()

[Transl_elong_EF1A_euk/arc (IPR004539), TRAFAC_GTPase_members (IPR050100)]    725
[TRAFAC_GTPase_members (IPR050100)]                                             7
[]                                                                              3
[Transl_elong_EF1A_euk/arc (IPR004539)]                                         2
Name: count, dtype: int64

In words: Out of 737 genes which this guide targets, there are 725 proteins where this guide cuts in 
both these two families [Transl_elong_EF1A_euk/arc (IPR004539), TRAFAC_GTPase_members (IPR050100)].
There are 7 proteins which this guide cuts in the TRAFAC_GTPase_members (IPR050100) family, and
2 more that are cut in Transl_elong_EF1A_euk/arc (IPR004539) family.

In [163]:
interpro[interpro['targeting_guide'] == 'GAGGCTGGTATCTCCAAGGA'].apply(find_target_superfamilies, axis=1).value_counts()

[P-loop_NTPase (IPR027417)]    735
[]                               2
Name: count, dtype: int64

In [164]:
interpro[interpro['targeting_guide'] == 'GAGGCTGGTATCTCCAAGGA'].apply(find_target_conserved_sites, axis=1).value_counts()

[]    737
Name: count, dtype: int64

In [165]:
interpro[interpro['targeting_guide'] == 'GAGGCTGGTATCTCCAAGGA'].apply(find_target_active_sites, axis=1).value_counts()

[]    737
Name: count, dtype: int64

In [167]:
interpro[interpro['targeting_guide'] == 'GAGGCTGGTATCTCCAAGGA'].apply(find_target_binding_sites, axis=1).value_counts()

[]    737
Name: count, dtype: int64

In [168]:
interpro[interpro['targeting_guide'] == 'GAGGCTGGTATCTCCAAGGA'].apply(find_target_other_features, axis=1).value_counts()

[EF1_alpha (cd01883)]                                                     709
[]                                                                         17
[Non cytoplasmic domain (NON_CYTOPLASMIC_DOMAIN), EF1_alpha (cd01883)]      5
[EF1_alpha (cd01883), Cytoplasmic domain (CYTOPLASMIC_DOMAIN)]              4
[Spurious_ORF_225 (ANF00225)]                                               2
Name: count, dtype: int64